packages/app/src/domain/vaccine/logic/get-vaccine-coverage-display-values.ts

```diff
@@ -0,0 +1,74 @@
+import { assert } from '@corona-dashboard/common';
+import { isDefined, isPresent } from 'ts-is-present';
+import { ChoroplethDataItem, MapType } from '~/components/choropleth/logic';
+import { replaceVariablesInText } from '~/utils/replace-variables-in-text';
+import { parseFullyVaccinatedPercentageLabel } from './parse-fully-vaccinated-percentage-label';
+
+export function getVaccineCoverageDisplayValues(
+  d: ChoroplethDataItem,
+  text: Record<string, Record<string, Record<string, string>>>,
+  map: MapType,
+  formatPercentage: (value: number) => string
+) {
+  const metrics = {
+    fully_vaccinated_percentage: '–',
+    has_one_shot_percentage: '–',
+  };
+
+  const getFormattedLabelOrValue = getLabelOrValueFormatter(
+    text,
+    map,
+    formatPercentage
+  );
+
+  if ('fully_vaccinated_percentage_label' in d) {
+    metrics.fully_vaccinated_percentage = getFormattedLabelOrValue(
+      d,
+      'fully_vaccinated_percentage',
+      'fully_vaccinated_percentage_label'
+    );
+  }
+
+  if ('has_one_shot_percentage_label' in d) {
+    metrics.has_one_shot_percentage = getFormattedLabelOrValue(
+      d,
+      'has_one_shot_percentage',
+      'has_one_shot_percentage_label'
+    );
+  }
+
+  return metrics;
+}
+
+function getLabelOrValueFormatter(
+  text: Record<string, Record<string, Record<string, string>>>,
+  map: MapType,
+  formatPercentage: (value: number) => string
+) {
+  return function <T extends ChoroplethDataItem>(
+    data: T,
+    property: keyof T,
+    labelKey: keyof T
+  ) {
+    const parsedValue = isPresent(data[labelKey])
+      ? parseFullyVaccinatedPercentageLabel(data[labelKey] as unknown as string)
+      : (data[property] as unknown as undefined | number | null) || null;
+
+    if (typeof parsedValue === 'number') {
+      return formatPercentage(parsedValue) + '%';
+    } else if (parsedValue !== null && 'sign' in parsedValue) {
```

Comment:
```suggestion
    } else if (isPresent(parsedValue) && 'sign' in parsedValue) {
```
From what I understand, in theory `parsedValue` can also be undefined, so even with the null check, this would lead to a runtime error.

Reply:
It should become `null` if the metric is undefined in the data because of the line you mentioned above, but might be good to use isPresent for consistency as well.

Reply:
Ah, yes, you're right, it does default to null